staging/src/k8s.io/client-go/resolver/resolver.go

```diff
@@ -0,0 +1,327 @@
+type ResolverOption func(*resolver) *resolver
+
+func WithTimeout(timeout time.Duration) func(*resolver) {
+	return func(r *resolver) {
+		r.timeout = timeout
+	}
+}
+
+// resolver to store API server IP addresses
+type resolver struct {
+	mu        sync.Mutex
+	cache     []net.IP  // API server IP addresses
+	timestamp time.Time // store the time the resolved is contacted once the cache has already been populated
+	host      string    // API server configured hostname
+	port      string    // API server configured port
+
+	// time after the cache entries are considered stale and pruned
+	timeout time.Duration
+	// RESTclient configuration
+	client *rest.RESTClient
+}
+
+// NewResolver returns an in memory net.Resolver that resolves the API server
+// Host name with the addresses obtained from the API server published Endpoints
+// resources.
+// The resolver polls periodically the API server to refresh the local cache.
+// The resolver will fall back to the default golang resolver if:
+// - Is not able to obtain the API server Endpoints.
+// - The configured API server host name is not resolvable via DNS, per example,
+//   is not an IP address or is resolved via /etc/hosts.
+// - The configured API server URL has a different port
+//   than the one used in the Endpoints.
+func NewResolver(ctx context.Context, c *rest.Config, options ...ResolverOption) (*net.Resolver, error) {
+	config := *c
+	if err := setConfigDefaults(&config); err != nil {
+		return nil, err
+	}
+
+	host, port, err := getHostPort(config.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	if netutils.ParseIPSloppy(host) != nil {
+		return net.DefaultResolver, fmt.Errorf("APIServerResolver only works for domain names")
+	}
+
+	// defaulting
+	r := &resolver{
+		host:    host,
+		port:    port,
+		timeout: 100 * time.Second,
+	}
+
+	// options
+	for _, o := range options {
+		o(r)
+	}
+
+	f := &dns.MemResolver{
+		LookupIP: func(ctx context.Context, network, host string) ([]net.IP, error) {
+			return r.lookupIP(ctx, network, host)
+		},
+	}
+
+	resolver := dns.NewMemoryResolver(f)
+	config.Resolver = resolver
+
+	client, err := rest.RESTClientFor(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	r.client = client
+	// Initialize cache and close idle connections so next connections goes
+	// directly to one of the API server IPs, this is useful for cases
+	// that use a load balancer for bootstrapping
+	r.refreshCache(ctx)
+	utilnet.CloseIdleConnectionsFor(r.client.Client.Transport)
+	return resolver, nil
+}
+
+func setConfigDefaults(config *rest.Config) error {
+	gv := v1.SchemeGroupVersion
+	config.GroupVersion = &gv
+	config.APIPath = "/api"
+	config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+
+	if config.UserAgent == "" {
+		config.UserAgent = rest.DefaultKubernetesUserAgent()
+	}
+
+	return nil
+}
+
+func (r *resolver) lookupIP(ctx context.Context, network, host string) ([]net.IP, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
```

Comment:
worth a RWMutex?

Reply:
There are several open issue on golang/go about the RWMutex scalability. 
wojtek confirmed that and advised me to use only if it was clear the benefit, I don't know if it applies here since lookups in flight are capped by the golang stdlib